### PR TITLE
Spatial_Engine:  Descriptions for Methods 

### LINKED ISSUE #1595: Spatial_Engine: Descriptions for methods
#### Broken rules:

There are no descriptions

#### Suggestions to restore compliance:

Add descriptions

### PR DESCRIPTION
**Pull request by kThorsager**
### Issues addressed by this PR

Closes #1595
Adds descriptions for all Spatial_Engine methods

### Test files
Not needed
### Additional comments

### CODE REVIEW

File: Spatial_Engine/Query/ElementVertices.cs [hunk 1]
```diff
@@ -52,6 +56,9 @@ public static List<Point> ElementVertices(this IElement1D element1D)
         /****            IElement2D            ****/
         /******************************************/
 
+        [Description("Queries the vertecies from a IElement2D. They are the defining element curves discontinuity points.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/ElementVertices.cs]:
```suggestion
        [Description("Returns the discontinuity points from the defining ICurves of the IElement2D.")]
```

File: Spatial_Engine/Query/ElementVertices.cs [hunk 2]
```diff
@@ -66,6 +73,9 @@ public static List<Point> ElementVertices(this IElement2D element2D)
 
         /******************************************/
 
+        [Description("Queries the vertecies from a collection of IElements. They are the defining element curves discontinuity points.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/ElementVertices.cs]:
```suggestion
        [Description("Returns the discontinuity points from the defining ICurves of the IElements.")]
```

File: Spatial_Engine/Query/ElementVertices.cs [hunk 3]
```diff
@@ -81,6 +91,9 @@ public static List<Point> ElementVertices(this IEnumerable<IElement> elements)
         /**** Public Methods - Interfaces      ****/
         /******************************************/
 
+        [Description("Queries the vertecies from a IElement. They are the defining element curves discontinuity points.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/ElementVertices.cs]:
```suggestion
        [Description("Returns the discontinuity points from the defining ICurves of the IElement.")]
```

File: Spatial_Engine/Query/IsSelfIntersecting.cs [hunk 4]
```diff
@@ -34,6 +37,10 @@ public static partial class Query
         /****            IElement1D            ****/
         /******************************************/
 
+        [Description("Returns if the one dimensional representation of the IElement1D is closer to itself than the tolerance at any two points.")]
+        [Input("element1D", "The IElement1D to evaluate self intersections from.")]
+        [Input("tolerance", "Minimum distance to be considered intersecting.", typeof(Length))]
+        [Output("o", "A boolean which is true if the IElement1Ds curve is self intersecting.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/IsSelfIntersecting.cs]: not sure I get the "o"?

File: Spatial_Engine/Query/IsSelfIntersecting.cs [hunk 5]
```diff
@@ -44,6 +51,10 @@ public static bool IsSelfIntersecting(this IElement1D element1D, double toleranc
         /****            IElement2D            ****/
         /******************************************/
 
+        [Description("Returns if any of the element curves of the IElement2D is closer to itself than the tolerance at any two points. Does not check for intersections between different curves.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/IsSelfIntersecting.cs]:
```suggestion
        [Description("Returns if any of the element curves of the IElement2D is closer to itself than the tolerance at any two points. Does not check for intersections between external and internal curves, or between different internal curves.")]
```

File: Spatial_Engine/Query/IsSelfIntersecting.cs [hunk 6]
```diff
@@ -44,6 +51,10 @@ public static bool IsSelfIntersecting(this IElement1D element1D, double toleranc
...
+        [Input("element2D", "The IElement2D which curves are to be evaluated for self intersection.")]
+        [Input("tolerance", "Minimum distance to be considered intersecting.", typeof(Length))]
+        [Output("o", "A boolean which is true if any of the IElement2Ds element curves are self intersecting.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/IsSelfIntersecting.cs]: "o" ?

File: Spatial_Engine/Query/Normal.cs [hunk 7]
```diff
@@ -33,6 +35,9 @@ public static partial class Query
         /****            IElement2D            ****/
         /******************************************/
 
+        [Description("Returns the normal of the external outline curve.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/Normal.cs]:
```suggestion
        [Description("Returns the normal based on the external outline curve.")]
```

File: Spatial_Engine/Query/Normal.cs [hunk 8]
```diff
@@ -33,6 +35,9 @@ public static partial class Query
...
+        [Input("element2D", "The IElement2D to get the normal from.")]
+        [Output("n", "The normal to the IElement2D which is perpendicular to its plane and oriented according to the rigth hand rule in relation to the outline curve.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/Normal.cs]: Really good description, but would maybe move some of this text from the output to the description of the method instead.

File: Spatial_Engine/Query/Centroid.cs [hunk 9]
```diff
@@ -45,6 +50,9 @@ public static Point Centroid(this IElement1D element1D)
         /****            IElement2D            ****/
         /******************************************/
 
+        [Description("Queries the centre of weight for a IElement2Ds surface representation.")]
```
[the remark this answers is not in the document]
Comment (kThorsager) [reply Spatial_Engine/Query/Centroid.cs]: Said it like this mostly as IElement1Ds descriptions are about centre of weight and wanted to keep some consistency.
agree to this but would like some suggestions on some equivalent word for the IElement1D one

File: Spatial_Engine/Query/IsSelfIntersecting.cs (hunk 4 above)
Comment (kThorsager) [reply Spatial_Engine/Query/IsSelfIntersecting.cs]: output, seemed silly to write `isSelfIntersecting` again

File: Spatial_Engine/Query/IsSelfIntersecting.cs (hunk 6 above)
Comment (kThorsager) [reply Spatial_Engine/Query/IsSelfIntersecting.cs]: same ^

File: Spatial_Engine/Query/Centroid.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Spatial_Engine/Query/Centroid.cs]: The one for `IElement1D` is not implemented from what I can see, so would rather consider stripping down that description all in all until we have a method in place for it, I think, and maybe add a `[NotImplemented]` tag.

I think the comment in that method is basically refering to centre of area. I am pretty sure it says "centre of weight" as opposed to "average control point", not as opposed to "centre of area".

File: Spatial_Engine/Query/Centroid.cs [hunk 10]
```diff
@@ -34,6 +36,9 @@ public static partial class Query
         /****            IElement1D            ****/
         /******************************************/
 
+        [Description("Queries the centre of weight for a IElement1Ds ICurve representation.")]
```
Comment (IsakNaslundBh) [inline Spatial_Engine/Query/Centroid.cs]:
```suggestion
        [NotImplemented]
        [Description("Queries the centre of weight for a IElement1Ds ICurve representation.")]
```

File: Spatial_Engine/Query/IsSelfIntersecting.cs (hunk 4 above)
Comment (al-fisher) [reply Spatial_Engine/Query/IsSelfIntersecting.cs]: I think for consistency of approach let's keep the full output name `isSelfIntersecting`.

Actually agree on the `OutputAttribute` there is duplication of information - forced from the format of the constructor requiring `name` & `description`  
``` C#
OutputAttribute(string name, string description, Type quantity = null)`
```

One to review more generally, but for now for clarity let us simply stick with full output name.
@IsakNaslundBh @kThorsager sound OK?

File: Spatial_Engine/Query/IsSelfIntersecting.cs (hunk 6 above)
Comment (al-fisher) [reply Spatial_Engine/Query/IsSelfIntersecting.cs]: See above ^ 😄

File: Spatial_Engine/Query/IsSelfIntersecting.cs (hunk 4 above)
Comment (kThorsager) [reply Spatial_Engine/Query/IsSelfIntersecting.cs]: Changed it and am OK, but should perhaps waited for happiness confirmation from @IsakNaslundBh first

Comment (al-fisher) [reply Spatial_Engine/Query/IsSelfIntersecting.cs]: Sorry - 🙈 just realised we have the constructor _already_ for OutputAttribute that takes a single string for description only. 👍 

``` C#

public OutputAttribute(string description)

```

@kThorsager  suggest you use this instead!

Comment (kThorsager) [reply Spatial_Engine/Query/IsSelfIntersecting.cs]: oh, when is the general case for not specifying a name?

**Comment by IsakNaslundBh** [comment]
Have checked the compliance issues, and it is nothing that needs to be resolved in this PR.

Will be fixed in https://github.com/BHoM/BHoM_Engine/issues/1572

**Pull request merged**